crates/control/src/motion/motor_command_optimizer.rs

```diff
@@ -0,0 +1,111 @@
+use color_eyre::Result;
+use context_attribute::context;
+use framework::{AdditionalOutput, MainOutput};
+use types::{
+    joints::{Joints, JointsCommand},
+    parameters::MotorCommandOptimizerParameters,
+    sensor_data::SensorData,
+};
+
+pub struct MotorCommandOptimizer {
+    position_offset: Joints<f32>,
+    is_resetting: bool,
+}
+
+#[context]
+pub struct CreationContext {}
+
+#[context]
+pub struct CycleContext {
+    pub motor_commands: Input<JointsCommand<f32>, "motor_commands">,
+    pub sensor_data: Input<SensorData, "sensor_data">,
+
+    pub params: Parameter<MotorCommandOptimizerParameters, "motor_command_optimizer_params">,
+
+    pub squared_position_offset_sum:
+        AdditionalOutput<f32, "motor_position_optimization_offset_squared_sum">,
+    pub position_offset: AdditionalOutput<Joints<f32>, "motor_position_optimization_offset">,
+}
+
+#[context]
+#[derive(Default)]
+pub struct MainOutputs {
+    pub optimized_motor_commands: MainOutput<JointsCommand<f32>>,
+}
+
+impl MotorCommandOptimizer {
+    pub fn new(_context: CreationContext) -> Result<Self> {
+        Ok(Self {
+            position_offset: Joints::default(),
+            is_resetting: false,
+        })
+    }
+
+    pub fn cycle(&mut self, mut context: CycleContext) -> Result<MainOutputs> {
+        let currents = context.sensor_data.currents;
+        let commands = *context.motor_commands;
+        let params = context.params;
+
+        let squared_position_offset_sum: f32 = self
+            .position_offset
+            .as_vec()
+            .into_iter()
+            .flatten()
+            .map(|position| position.powf(2.0))
+            .sum();
+
+        if squared_position_offset_sum > params.offset_reset_threshold {
+            self.is_resetting = true;
+        }
+
+        if self.is_resetting {
+            self.position_offset = self.position_offset / params.offset_reset_speed;
+
+            if squared_position_offset_sum
+                < params.offset_reset_threshold / params.offset_reset_offset
+            {
+                self.is_resetting = false;
+            }
+        }
+
+        let maximal_current = currents.as_vec().into_iter().flatten().fold(0.0, f32::max);
+
+        let position_offset = params
+            .optimization_direction
+            .as_vec()
+            .into_iter()
+            .flatten()
+            .zip(currents.as_vec().into_iter().flatten())
+            .map(|(correction_direction, current)| {
+                if current == maximal_current {
+                    params.optimization_speed * correction_direction
+                } else {
+                    0.0
+                }
+            });
+
+        if maximal_current >= params.optimization_current_threshold && !self.is_resetting {
+            self.position_offset = self.position_offset + Joints::from_iter(position_offset);
+        }
+
+        let mut optimized_stiffnesses = commands.stiffnesses;
```

Comment:
do not use `mut`, but instead construct the `Joints` with `Joints { a: 0.0, b: 0.0, ..commands.stiffnesses}`

Reply:
I chose this because it reduces the lines of code by a factor of 7:

`        let mut optimized_stiffnesses = commands.stiffnesses;
        optimized_stiffnesses.left_arm.hand = 0.0;
        optimized_stiffnesses.right_arm.hand = 0.0;

        let optimized_stiffnesses_2 = Joints {
            head: commands.stiffnesses.head,
            left_arm: ArmJoints {
                shoulder_pitch: commands.stiffnesses.left_arm.shoulder_pitch,
                shoulder_roll: todo!(),
                elbow_yaw: todo!(),
                elbow_roll: todo!(),
                wrist_yaw: todo!(),
                hand: todo!(),
            },
            right_arm: ArmJoints {
                shoulder_pitch: todo!(),
                shoulder_roll: todo!(),
                elbow_yaw: todo!(),
                elbow_roll: todo!(),
                wrist_yaw: todo!(),
                hand: todo!(),
            },
            left_leg: commands.stiffnesses.left_leg,
            right_leg: commands.stiffnesses.right_leg,
        };
`

Reply:
Okay hold on, got it, this is actually awesome 😆
```
        let optimized_stiffnesses = Joints {
            left_arm: ArmJoints {
                hand: 0.0,
                ..commands.stiffnesses.left_arm
            },
            right_arm: ArmJoints {
                hand: 0.0,
                ..commands.stiffnesses.right_arm
            },
            ..commands.stiffnesses
        };
```